cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -1364,15 +1364,41 @@ impl fmt::Debug for Inst {
     }
 }
 
+/// Is the given (AluRmiROpcode, src, dst) combination causing a redefinition of the input operand,
+/// instead of an actual use?
+fn alu_rmi_r_is_def(src: &RegMemImm, dst: Writable<Reg>, op: AluRmiROpcode) -> bool {
+    op == AluRmiROpcode::Xor
+        && if let RegMemImm::Reg { reg: src } = src {
+            *src == dst.to_reg()
+        } else {
+            false
+        }
+}
+
+/// Is the given (SseOpcode, src, dst) combination causing a redefinition of the input operand,
+/// instead of an actual use?
+fn xmm_rm_r_is_def(src: &RegMem, dst: Writable<Reg>, op: SseOpcode) -> bool {
+    (op == SseOpcode::Xorps || op == SseOpcode::Xorpd)
+        && if let RegMem::Reg { reg: src } = src {
+            *src == dst.to_reg()
+        } else {
+            false
+        }
+}
+
 fn x64_get_regs(inst: &Inst, collector: &mut RegUsageCollector) {
     // This is a bit subtle. If some register is in the modified set, then it may not be in either
     // the use or def sets. However, enforcing that directly is somewhat difficult. Instead,
     // regalloc.rs will "fix" this for us by removing the the modified set from the use and def
     // sets.
     match inst {
-        Inst::Alu_RMI_R { src, dst, .. } => {
-            src.get_regs_as_uses(collector);
-            collector.add_mod(*dst);
+        Inst::Alu_RMI_R { src, dst, op, .. } => {
+            if alu_rmi_r_is_def(src, *dst, *op) {
```

Comment:
We should have a comment here describing what's going on, perhaps similar to my suggestion in @abrown's patch [here](https://github.com/bytecodealliance/wasmtime/pull/2060#discussion_r460293433).